targets/ublk.nfs.cpp

```diff
@@ -0,0 +1,383 @@
+                            iod->nr_sectors * 512, iod->start_sector * 512,
+                            rw_async_cb, cb_data) < 0) {
+                ublk_err("Failed to read from nfs file. %s\n", nfs_get_error(nfs_data->nfs));
+                free(cb_data);
+                return -ENOMEM;
+        }
+
+        return 0;
+}
+
+static int nfs_tgt_write(const struct ublksrv_queue *q,
+                         const struct ublksrv_io_desc *iod, int tag)
+{
+        const struct ublksrv_dev *dev = q->dev;
+        struct nfs_tgt_data *nfs_data = (struct nfs_tgt_data *)dev->tgt.tgt_data;
+        nfs_cb_data_t *cb_data;
+
+        if ((iod->nr_sectors + iod->start_sector) * 512 > nfs_data->capacity) {
+                return -EINVAL;
+        }
+        
+        cb_data = (nfs_cb_data_t *)malloc(sizeof(*cb_data));
+        if (cb_data == NULL) {
+                ublk_err("Malloc failed in nfs_tgt_read");
+                return -ENOMEM;
+        }
+        cb_data->q = q;
+        cb_data->tag = tag;
+        
+        if (nfs_pwrite_async(nfs_data->nfs, nfs_data->nfsfh,
+                             (void *)iod->addr,
+                             iod->nr_sectors * 512, iod->start_sector * 512,
+                             rw_async_cb, cb_data) < 0) {
+                ublk_err("Failed to write to nfs file. %s\n", nfs_get_error(nfs_data->nfs));
+                free(cb_data);
+                return -ENOMEM;
+        }
+
+        return 0;
+}
+
+static int nfs_tgt_flush(const struct ublksrv_queue *q,
+                         const struct ublksrv_io_desc *iod, int tag)
+{
+        const struct ublksrv_dev *dev = q->dev;
+        struct nfs_tgt_data *nfs_data = (struct nfs_tgt_data *)dev->tgt.tgt_data;
+        nfs_cb_data_t *cb_data;
+
+        cb_data = (nfs_cb_data_t *)malloc(sizeof(*cb_data));
+        if (cb_data == NULL) {
+                ublk_err("Malloc failed in nfs_tgt_read");
+                return -ENOMEM;
+        }
+        cb_data->q = q;
+        cb_data->tag = tag;
+
+        if (nfs_fsync_async(nfs_data->nfs, nfs_data->nfsfh,
+                            rw_async_cb, cb_data) < 0) {
+                ublk_err("Failed to fsync nfs file. %s\n", nfs_get_error(nfs_data->nfs));
+                free(cb_data);
+                return -ENOMEM;
+        }
+
+        return 0;
+}
+
+static int nfs_handle_io_async(const struct ublksrv_queue *q,
+		const struct ublk_io_data *data)
+{
+        const struct ublksrv_io_desc *iod = data->iod;
+        unsigned ublk_op = ublksrv_get_op(iod);
+        int ret = -ENOTSUP;
+        
+        switch (ublk_op) {
+        case UBLK_IO_OP_READ:
+                ret = nfs_tgt_read(q, iod, data->tag);
+                break;
+        case UBLK_IO_OP_WRITE:
+                ret = nfs_tgt_write(q, iod, data->tag);
+                break;
+        case UBLK_IO_OP_DISCARD:
+                ublk_err("UBLK_IO_OP_DISCARD is not supported");
+                break;
```

Comment:
ublk is complete async io handling, here not supported commands have to be completed explicitly by ublksrv_complete_io(-ENOTSUP), otherwise you still may hang forever if kernel sends discard/write zero to us.

Reply:
looks you already do it.